fixes #5130 - Added warning when apipie cache is missing or outdated

lzap opened the pull request:
This adds one small check during startup when in development mode. The test
takes less than 0.050 seconds currently and check is apipie cache:

a) is present (if not - WARNING)

b) is outdated (if it is - WARNING)

Particularly state (b) is quite dangerous, I can see myself spending two hours
debugging a API code with outdated cache.

If you like this, I will make an issue for it.

lzap committed:
fixes #5130 - Added warning when apipie cache is missing or outdated

domcleal commented on config/initializers/apipie.rb:
```diff
@@ -1,3 +1,19 @@
+  cache_name = "public/apipie-cache/apidoc.json"
+  if File.exists? cache_name
+    target = max = File.mtime(cache_name)
+    Find.find('app/controllers/api') { |e| t = File.mtime(e); max = t if t > max }
```
Will this work with engines?

lzap replied:
> Will this work with engines?

No, apparently.

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

domcleal replied:
Could it please? :)

Maybe we can iterate over Rails.application.railties.engines?

lzap replied:
> Could it please? :)

Sure, just say it directly :-)

> Maybe we can iterate over Rails.application.railties.engines?

Oh thanks, easy enough with this information for me now :)

Which brings me to the following question on @mbacovsky:

I have just enabled a plugin with api extension (discovery from this
branch - https://github.com/theforeman/foreman_discovery/pull/61) and
regenerated apipie:cache, but in the generated JSON I do not see a
single string "discovery" or "discovers".

Does apipie:cache also extract info from Foreman Plugins?

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

domcleal commented on config/initializers/apipie.rb:
```diff
@@ -1,3 +1,19 @@
+      puts "WARNING: API controllers newer than Apipie cache! Run 'rake apipie:cache' to regenerate cache."
+    end
+  else
+    puts "WARNING: Apipie cache not present yet. Run 'rake apipie:cache' to regenerate cache."
```
Is this a duplicate warning to the one after the apipie configure call?

lzap replied:
> Is this a duplicate warning to the one after the apipie configure call?

Sorry what? It is also important to inform the dev about missing cache I
think.

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

domcleal replied:
There's another piece of logging after the apipie call, won't they both fire?
https://github.com/lzap/foreman/blob/warning-apipie/config/initializers/apipie.rb#L40-L43

mbacovsky replied:
yes, they will. It also may be good to add warning that the cache does not exist even in production mode to avoid running with use_cache and no cache ready.

If we move these cache checks after the configuration part we could use Apipie.configuration.use_cache in the conditions:

```
use_cache && cache outdated or missing -> warning or error(!)
!use_cache -> warning it could be slow to use CLI
```

As a bonus we could then replace cache_name with 

```
File.join(Apipie.configuration.cache_dir, Apipie.configuration.doc_base_url+'.json')
```

There is also list of all controllers that apipie is interested in in `Apipie.configuration.api_controllers_matcher` But I'm not sure if we can move this checks after all engines initializers are finished. That way we would have there all the relevant paths stored.

Makes sense?

domcleal commented:
I like the idea, do you agree @mbacovsky?

mbacovsky commented:
Sure, it is useful addition. 
BTW: Wouldn't it be better to test for !Rails.env.production?

lzap commented:
> Sure, it is useful addition. 
> BTW: Wouldn't it be better to test for !Rails.env.production?

Yeah ok.

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

lzap commented:
Pushed.

Note I am using puts instead of warn becaue I do not want these messages to be lost in the development.log. We want to warn users. In production we should never seen any of these.

mbacovsky commented:
I like that, it works nice. Not sure how to handle different name for rake command in production (foreman-rake). Do we have some function returning the actual rake name? Other than that it seems ready to go.

lzap commented:
Those warnings should not be issued in production. I guess we do not
want foreman-rake.

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

mbacovsky commented:
@lzap especially "Apipie cache enabled but not present yet. Run 'rake apipie:cache' to speed up API calls." is very likely to appear in prod.

lzap commented:
> especially "Apipie cache enabled but not present yet. Run 'rake apipie:cache' to speed up API calls." is very likely to appear in prod.

Well, I assume we will run apipie:cache for packages, so it should not.

On the other hand, one could change a controller (e.g. using a patch or
something).

But stdout is unlikely to be visible anyway, you start foreman with:

service httpd restart

Thus this message will not be visible.

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

domcleal commented:
Yeah, the packages should take care of running this after each update, so I think it's OK.  However users may see it when either running foreman-rake commands or in log files, so I'd like to fix the foreman-rake versus rake text first.  I'd simply say print "foreman-rake" when `Rails.env.production?` else print "rake".

Once that's fixed, ready to merge.

lzap commented:
Sorry for the delay, missed this one.

Pushed the change that @orrabin filed here https://github.com/lzap/foreman/pull/6

Here is the change from the last version:

``` diff
 # check apipie cache in dev mode
+rake_name = Rails.env.production? ? 'forman-rake apipie:cache' : 'rake apipie:cache'
 if Apipie.configuration.use_cache
   cache_name = File.join(Apipie.configuration.cache_dir, Apipie.configuration.doc_base_url + '.json')
   if File.exists? cache_name
@@ -30,13 +31,13 @@ if Apipie.configuration.use_cache
       Find.find(path) { |e| t = File.mtime(e); max = t if t > max } if File.exists?(path)
     end
     if max > target
-      puts "API controllers newer than Apipie cache! Run 'rake apipie:cache' to regenerate cache."
+      puts "API controllers newer than Apipie cache! Run '#{rake_name}' to regenerate cache."
     end
   else
-    puts "Apipie cache enabled but not present yet. Run 'rake apipie:cache' to speed up API calls."
+    puts "Apipie cache enabled but not present yet. Run '#{rake_name}' to speed up API calls."
   end
 else
-  puts "The Apipie cache is turned off. Enable it and run 'rake apipie:cache' to speed up API calls."
+  puts "The Apipie cache is turned off. Enable it and run '#{rake_name}' to speed up API calls."
 end
```

isratrade commented:
@lzap, I ran locally 

```
$ rake apipie:cache
API controllers newer than Apipie cache! Run 'rake apipie:cache' to regenerate cache.
2014-09-15 14:36:41 +0300 | Started
2014-09-15 14:36:48 +0300 | Documents loaded...
2014-09-15 14:36:48 +0300 | Processing docs for 
```

I expected if I'm running `rake apipie:cache` then I wouldn't the message to run it.

lzap commented:
Right, rebased, added check for command line option.

``` diff
       Find.find(path) { |e| t = File.mtime(e); max = t if t > max } if File.exists?(path)
     end
-    if max > target
+    binding.pry
+    if $ARGV.first != "apipie:cache" && max > target
       puts "API controllers newer than Apipie cache! Run '#{rake_name}' to regenerate cache."
     end
   else
```

lzap commented:
Removed the `binding.pry` line, re-[test](Jenkins killed the job for unknown reason). @isratrade ?

lzap commented:
[test] because not sure if Jenkins would pick that up

dLobatog commented on config/initializers/apipie.rb:
```diff
@@ -17,9 +19,25 @@
-  warn "The Apipie cache is turned off.\n" \
-    "  To improve performance of your API clients turn it on by running 'rake apipie:cache' and restart the server."
+# check apipie cache in dev mode
+rake_name = Rails.env.production? ? 'forman-rake apipie:cache' : 'rake apipie:cache'
```
Typo `forman`

dLobatog commented:
Thanks @lzap , looks useful for people using apipie cache. 
Only one concern aside from the typo: please check if `foreman-rake` is available, and if it's not, then use `rake apipie:cache`, the fact that it's running on production doesn't really mean the owner is using foreman-rake, it can be installed from source. 

Another problem I found is that $ARGV might be empty.

```
daniel::localhost { ~/workspace/cleanforeman }-> RAILS_ENV=production rails s
=> Booting Thin
=> Rails 3.2.18 application starting in production on http://0.0.0.0:3000
=> Call with -d to detach
=> Ctrl-C to shutdown server
Exiting
/home/daniel/workspace/cleanforeman/config/initializers/apipie.rb:33:in `<top (required)>': undefined method `first' for nil:NilClass (NoMethodError)
```

lzap commented:
All fixed, rebased.

## 

Later,
 Lukas #lzap Zapletal

dLobatog commented:
[test]

dLobatog commented:
Works well now, thanks @lzap . I'll merge when tests are :green_apple:

dLobatog commented:
Merged as https://github.com/theforeman/foreman/commit/a6cdb17c44ce2f1e5bf12a544ad2fa7d81e97032 , thanks @lzap !
